bridge/pkg/terra/watcher.go

```diff
@@ -0,0 +1,224 @@
+			tokenChain := gjson.Get(json, "result.events.from_contract\\.locked\\.token_chain.0")
+			tokenDecimals := gjson.Get(json, "result.events.from_contract\\.locked\\.token_decimals.0")
+			token := gjson.Get(json, "result.events.from_contract\\.locked\\.token.0")
+			sender := gjson.Get(json, "result.events.from_contract\\.locked\\.sender.0")
+			recipient := gjson.Get(json, "result.events.from_contract\\.locked\\.recipient.0")
+			amount := gjson.Get(json, "result.events.from_contract\\.locked\\.amount.0")
+			nonce := gjson.Get(json, "result.events.from_contract\\.locked\\.nonce.0")
+			txHash := gjson.Get(json, "result.events.tx\\.hash.0")
+
+			if targetChain.Exists() && tokenChain.Exists() && tokenDecimals.Exists() && token.Exists() &&
+				sender.Exists() && recipient.Exists() && amount.Exists() && amount.Exists() && nonce.Exists() && txHash.Exists() {
+
+				logger.Info("Token lock detected on Terra")
+				logger.Info("\ttxHash: ", zap.Any("txHash", txHash))
+				logger.Info("\ttargetChain: ", zap.Any("targetChain", targetChain))
+				logger.Info("\ttokenChain: ", zap.Any("tokenChain", tokenChain))
+				logger.Info("\ttokenDecimals: ", zap.Any("tokenDecimals", tokenDecimals))
+				logger.Info("\ttoken: ", zap.Any("token", token))
+				logger.Info("\tsender: ", zap.Any("sender", sender))
+				logger.Info("\trecipient: ", zap.Any("recipient", recipient))
+				logger.Info("\tamount: ", zap.Any("amount", amount))
+				logger.Info("\tnonce: ", zap.Any("nonce", nonce))
+
+				senderAddress, err := StringToAddress(sender.String())
+				if err != nil {
+					logger.Error("cannod decode hex ", zap.Any("value", sender.String()))
+					continue
+				}
+				recipientAddress, err := StringToAddress(recipient.String())
+				if err != nil {
+					logger.Error("cannod decode hex ", zap.Any("value", recipient.String()))
+					continue
+				}
+				tokenAddress, err := StringToAddress(token.String())
+				if err != nil {
+					logger.Error("cannod decode hex ", zap.Any("value", token.String()))
+					continue
+				}
+				txHashValue, err := StringToHash(txHash.String())
+				if err != nil {
+					logger.Error("cannod decode hex ", zap.Any("value", txHash.String()))
+					continue
+				}
+				lock := &common.ChainLock{
+					TxHash:        txHashValue,
+					Timestamp:     time.Now(), // No timestamp available, consider adding it into transaction logs or request additionally from the blockchain
+					Nonce:         uint32(nonce.Uint()),
+					SourceAddress: senderAddress,
+					TargetAddress: recipientAddress,
+					SourceChain:   vaa.ChainIDTerra,
+					TargetChain:   vaa.ChainID(uint8(targetChain.Uint())),
+					TokenChain:    vaa.ChainID(uint8(tokenChain.Uint())),
+					TokenAddress:  tokenAddress,
+					TokenDecimals: uint8(tokenDecimals.Uint()),
+					Amount:        new(big.Int).SetUint64(amount.Uint()),
+				}
+				e.lockChan <- lock
+			}
+
+			// Query and report guardian set status
```

Comment:
Why is this done in the websocket receive loop?

Reply:
In this loop we check for all actions on the wormhole smart contract. So on each action we query guardian set from the contract. But since ETH is the source of truth (as you stated in the other comment) there is no need to do this I guess.